src/S3/Transfer.php

```diff
@@ -340,7 +346,37 @@ private function getDownloadsIterator()
         return $this->source;
     }
 
-    private function upload($filename)
+    private function uploadAsNeeded($file)
+    {
+        if (!$this->overwrite) {
+            return $this->client
+                ->executeAsync($this->hasBeenUploadedCommand($file))
```

Comment:
There is already a doesObjectExist function built into the SDK as part of the S3ClientTrait which would be useful here to reduce code duplication.  This could be fewer lines of code by having it just a small addition to the upload function instead of it's own function:
if (!this->overwrite)  { 
    if ($this->client ->doesObjectExist($file)) {
        return
    }
}
//normal flow of logic continues

Reply:
It does indeed work better.

Reply:
I initially thought about using the "doesObjectExist" function. I think it depends on how we want to achieve concurrency :
* one option is to first filter files that need to be uploaded and then create upload promises only for these files. In that case, using the doesObjectExist would be the best solution
* the other option is to test existence concurently : for each files, we create a promise that will test AND upload as needed. I think this is faster, that's why I prefered this implementation.